arangod/Aql/ExecutionNode.cpp

```diff
@@ -1198,48 +1200,34 @@ RegisterId ExecutionNode::variableToRegisterOptionalId(Variable const* var) cons
   return RegisterPlan::MaxRegisterId;
 }
 
-bool ExecutionNode::isIncreaseDepth() const {
+bool ExecutionNode::isPassthrough() const {
   switch (getType()) {
     case ExecutionNode::ENUMERATE_COLLECTION:
     case ExecutionNode::INDEX:
     case ExecutionNode::ENUMERATE_LIST:
     case ExecutionNode::COLLECT:
 
-
     case ExecutionNode::TRAVERSAL:
     case ExecutionNode::SHORTEST_PATH:
     case ExecutionNode::K_SHORTEST_PATHS:
 
     case ExecutionNode::REMOTESINGLE:
     case ExecutionNode::ENUMERATE_IRESEARCH_VIEW:
     case ExecutionNode::MATERIALIZE:
-      return true;
+      return false;
 
     default:
-      return false;
+      return true;
   }
 }
 
 /// @brief creates corresponding ExecutionBlock
 std::unique_ptr<ExecutionBlock> SingletonNode::createBlock(
     ExecutionEngine& engine, std::unordered_map<ExecutionNode*, ExecutionBlock*> const&) const {
-  // number in == number out
-  // Other nodes get the nrInRegs from the previous node.
-  // That is why we do not use `calcRegsToKeep()`
-  RegisterId const nrRegs = getRegisterPlan()->nrRegs[getDepth()];
-
-  std::unordered_set<RegisterId> regsToKeep;
-  if (isInSubquery()) {
-    for (auto const& var : this->getVarsUsedLater()) {
-      auto val = variableToRegisterId(var);
-      if (val < nrRegs) {
-        auto rv = regsToKeep.insert(val);
-        TRI_ASSERT(rv.second);
-      }
-    }
-  }
 
+  RegisterId const nrRegs = getRegisterPlan()->nrRegs[getDepth()];
   auto regsToClear = getRegsToClear();
+  auto regsToKeep = calcRegsToKeep();
 
   auto registerInfos = RegisterInfos{make_shared_unordered_set(),
                                      make_shared_unordered_set(),
```

Comment:
Very good to get rid of this special case!

And at this point you can just call
```
auto registerInfos = createRegisterInfos(make_shared_unordered_set(), make_shared_unordered_set());
```
which calls `calcRegsToKeep` etc.